### ISSUE TITLE
Provide help for unsolved/unfinished exercises

### ISSUE THREAD

**Issue by henningsway** [Tazinho/Advanced-R-Solutions#228]
After working through the reviews provided by Hadley, we probably will have a couple of exercises where the final solution isn't yet entirely clear or satisfying yet.

We will collect these exercises here and would appreciate a second look or solution provided by Hadley.

#### 2. Names and Values
- [x] 4.2 Take the following list. Why is its size somewhat misleading? ...
  - Not sure of the answer really answers the question
  - I could imagine several ways how this question is meant. A bit of clarification would be helpful.

#### 3. Vectors
- [x] 1.1 How do you create scalars of type raw and complex? (See ?raw and ?complex)...
  - Could you explain what a raw vector is and for what it is used (maybe also how it is handled or similar). I think it would really be helpful for some of the audience (me in cluded)
- [x] 3.1 What sort of object does table() return? What is its type? What attributes does it have? How does ...
  - Here we write that the output of `table()` has the implicit class `array`? Is this really correct? The documentation states that that the implicit class is the type (so integer) if the object doesn't have a class (i.e. attr(x, "class") is NULL)
- [ ] Lists exercise 3: Compare and contrast c() and unlist() when combining a date and date-time into a single vector. ...
  - the combination of different date types via `c()` doesn't crash anymore in R 4.0

#### 5. Control Flow
- [x] 1.1 What type of vector does each of the following calls to `ifelse()` return? ...
  - I made some chages to this solution and would be happy if you'd just review it again.

#### 6. Functions
- [ ] 3.1 (optional) What important property of && makes x_ok() work?...
  - This answer still reads far too technical and is probably a bit off from the actual intended question.

- [ ] 5.5 We can use on.exit() to implement a simple version of capture.output(). ...
    - We changed a bit here and would be happy if you'ld have another look on this one
    - Especially we are not sure about the statements regarding the error handling and the robustness of the lapply(eval(substitute)) approach.

#### 10. Function factories
- [ ] (optional issue) 2.2 Base R contains two function factories, approxfun() and ecdf(). Read their documentation and ...
    - The text might benefit from using the correct mathematical terms for the initial x,y values and their combinations. In german we have the terms "Stützwerte", "Stützstellen" oder "Knotenpunkte (nodes)". I would probably mix these incorrectly in English...

#### 13. S3
- [ ] 5.1 How does [.Date support subclasses? How does it fail to support subclasses?
  - We don't fully understand this solution and would appreciate more context for motivation or an example for clarification
  - Are there any concrete implications for the user which result from the part about the internals? Maybe this would guide a bit on where to lay the focus while formulating this part.
  - We don't see why `mean()` is mentioned in the solution

#### 21. Translating R code
  - [ ] 2.2 Complete the DSL to support all the functions that plotmath supports. ...
    - We are still not that familiar with DSL programming and the solution for this exercise doesn't come to us easily. We thought about possibly choosing `to_math(x+y)`, `to_math(x %*% y)`, `to_math(bold(x))` and `to_math(sum(x[i], i==1, n))` as examples. Could we persuade you to provide the solution for this exercise?

#### 23. Measuring performance
- [x] 1.1 Profile the following function with `torture = TRUE`. What is surprising? ...
  - We don't really know how to improve/ finish this exercise + the current code takes a very long time to run. Could you take another look at this exercise?

#### 24. Improving performance
  - [ ] 1.3 List four functions (not just those in base R) that convert a string into a date time object. What are ...
    - `c(date_lt, date_ct)` and `c(date_ct, date_lt)` seems to be fixed (probably du to R version) so this part needs to be edited sowhow.
    - is it still ill advised to combine dates (we guess so, because of single dispatch). Do you have another good crashing example or would you just modify the solution a bit?
    - Unfortunately, this issue also comes up in the Vectors chapter for the 3rd exercise in the lists section ("Compare and contrast c() and unlist() when combining a date and date-time into a single vector.")
  - [ ] 2.3 Can you make a faster version of table() for the case of an input of two integer vectors with no ...
    - HW: are you sure you need fastmatch here? I think you should be able to work out the bin with a little algebra
      - We can only think of the following alternative which is more readable, but slower (maybe we can put it in comments to help the reader)

        ```
        m <- matrix(1:pr, nrow = a_l, dimnames = dn)
        bin <- m[matrix(as.character(c(a, b)), ncol = 2)]
        ```

    - So, we think we need another hint on how to use some algebra to calculate the bin (faster).
    - We also still need to use the speed improvement for chi.squared downstream.

### Rcpp
- [ ] 3.6 I think there is a bug in `union()` and `setdiff()`. See also #240

**Comment by Tazinho** [comment]
@hadley we have reviewed all the solutions and have also incorporated the comments from your reviews. For the above exercises however, we would need your support once again.

**Comment by hadley** [comment]
I'll do my best to get to it this week.

**Comment by Tazinho** [comment]
Thank you, Hadley

**Comment by hadley** [comment]
> Take the following list. Why is its size somewhat misleading?

Answer seems fine to me — what does it mean to measure the size of something that's already included in R? There's a more general question about what you want to know when you ask for the size of something — do you want to know how much data you'd need to send to communicate the object to someone else (e.g. serialize it), or do you want to know how much memory you'd free if you deleted it? 

> How do you create scalars of type raw and complex?

I'd say: There's no way to create individual raw and complex values — there's no built in syntax like there is for logicals, integers, doubles, and character vectors. Instead you have to create them by calling a function, either `as.raw()` or `rawToChar()` for raw vectors, or `+` or `complex()` in the case of complex numbers.

You might want to mention that you can create imaginary numbers with (e.g.) `1i`, but there's no way to create complex numbers without `+`.

> What sort of object does table() return? What is its type? What attributes does it have? 

I'd say it returns an integer vector with class "table" and dimensions (which makes it act like an array). It technically doesn't have an implicit class because it has an explicit class — this is one of the places where S3 isn't particularly well defined. You might want to emphasise that it behaves like an array by showing some subsetting.

> Compare and contrast c() and unlist() when combining a date and date-time into a single vector.

I think you can say: As of R 4.0.0, `c()` preserves the class of the first argument, so you that `c(x, y)` is not guaranteed to return the same type as `c(y, x)`. I'd mention that this wasn't always the case, and also point out that `c()` will strip the time zone from POSIXct, so some caution is recommended. (I'd recommend using something other than UTC as the time zone to make this more clear)

**Comment by hadley** [comment]
> What type of vector does each of the following calls to ifelse() return? 
  
LGTM 

> What important property of && makes x_ok() work?...

I'd just say: `&&` short-circuits which means that if the left-hand side is `FALSE` it doesn't evaluate the right-hand side (because it doesn't matter). Similarly, if the left-hand side of `||` is `TRUE` it doesn't evaluate the right-hand side.

> We can use on.exit() to implement a simple version of capture.output(). ...

I don't think that `capture.output()` is any more robust - both will terminate after the first error. I think the main difference is that `capture.output()` calls `print` , i.e. compare the output of these two calls:

```R
capture.output({1})
capture.output2({1})
```

Weirdly, this doesn't appear to be documented.

The `substitute()` machinery just lets you figure out whether or not the output was visible, e.g so that `capture.output(x <- 1)` correctly returns `character()`.

> Base R contains two function factories, approxfun() and ecdf(). Read their documentation and ..

The explanations seem fine to me. I'd just make them a bit more concrete with some examples, e.g.

```R
x <- rnorm(100)
y <- ecdf(x)
plot(y)
y(-1)

f <- approxfun(c(0, 1, 2), c(0, 10, 100))
f(0.5)
plot(f, xlim = c(0, 2))
```

>  How does [.Date support subclasses? How does it fail to support subclasses?

I think you can shrink the explanation of `oldClass()` a bit - just say that it's basically the same as `class()`, except that it doesn't return implicit classes, i.e. it's basically `attr(x, "class")` (looking at the C code that's exactly what it does, except that it also handles S4 objects).

So if `oldClass()` is "basically" class, you can rewrite `[.Date` to make the implementation more clear:

```R
`[.Date` <- function(x, ..., drop = TRUE) {
  out <- .Date(NextMethod("["), oldClass(x))
  class(out) <- class(x)
  out
}
```

So `[.Date` ensures that the output has the same class as in the input. But what about other attributes that a subclass might possess? They get lost:

```R
x <- structure(1:4, test = "test", class = c("myDate", "Date"))
attributes(x[1])
```

**Comment by hadley** [comment]
> Complete the DSL to support all the functions that plotmath supports. ..

I think it's fine to just pick a handful and implement them.

>  Profile the following function with torture = TRUE. What is surprising? 

I'd say to just scrap it — I can't get it to complete on R 4.0.0 either. I did manage to eventually get a profile in 3.5 (which was the current version of R at the time I wrote the book), but it wasn't particularly informative. Just say something like: "This question appears to be unanswerable, even by Hadley"

**Comment by hadley** [comment]
>  List four functions (not just those in base R) that convert a string into a date time object. What are 

I'd focus on `strptime()` (since it powers everything else), lubridate `ymd_hms()`, fasttime, and anytime. I'd suggest doing a benchmark, and then briefly summarising the types of date time input each accepts.

>  Can you make a faster version of table() for the case of an input of two integer vectors with no ...

Does this help?

```R
x <- matrix(1:20, nrow = 4)
(x - 1) %% 4 + 1
(x - 1) %/% 4 + 1
```

**Comment by Tazinho** [comment]
Closed in favour of #254
